compiler/src/dotty/tools/dotc/transform/PureStats.scala

```diff
@@ -0,0 +1,32 @@
+package dotty.tools.dotc
+package transform
+
+import ast.{Trees, tpd}
+import core._, core.Decorators._
+import MegaPhase._
+import Types._, Contexts._, Flags._, DenotTransformers._
+import Symbols._, StdNames._, Trees._
+
+object PureStats {
+  val name: String = "pureStats"
+}
+
+/** Remove pure statements in blocks */
+class PureStats extends MiniPhase {
+
+  import tpd._
+
+  override def phaseName: String = PureStats.name
+
+  override def runsAfter: Set[String] = Set(Erasure.name)
+
+  override def transformBlock(tree: Block)(implicit ctx: Context): Tree =
+    val stats = Trees.flatten(
+    tree.stats.mapConserve {
+      case Typed(Block(stats, expr), _) if isPureExpr(expr) => Thicket(stats)
+      case stat if !stat.symbol.isConstructor && isPureExpr(stat) => EmptyTree
+      case stat => stat
+    })
+    cpy.Block(tree)(stats, tree.expr)
```

Comment:
Since you're using `mapConserve` you may not need to call `Trees.flatten` and `cpy.Block`:
```scala
val stats = tree.stats.mapConserve { ... }
if stats eq tree.stats then tree
else cpy.Block(tree)(Trees.flatten(stats), tree.expr)
```